{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import collin_screen_analysis as sca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "b_top = pd.read_csv(\"Genome_Lipid_Free_hiSHH_Top5_combined_mageck_input.txt\", delimiter=\"\\t\")\n",
    "b_bot = pd.read_csv(\"Genome_Lipid_Free_hiSHH_Bot10_combined_mageck_input.txt\", delimiter=\"\\t\")\n",
    "c_guides = pd.read_csv(\"hiSHH_full_analysis_guide_counts.csv\")\n",
    "\n",
    "result = sca.create_mageck_input_files(c_guides, [\"Unsorted Population\", \"Top Sorted Population\", \"Bottom Sorted Population\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/collinschlager/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:2530: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "  obj = obj._drop_axis(labels, axis, level=level, errors=errors)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>condition</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Unsorted Population</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Top Sorted Population</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Bottom Sorted Population</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate</th>\n",
       "      <th>Rep1</th>\n",
       "      <th>Rep2</th>\n",
       "      <th>Rep1</th>\n",
       "      <th>Rep2</th>\n",
       "      <th>Rep1</th>\n",
       "      <th>Rep2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgRNA Target Sequence</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAAAAAAAAAATGCATTCT</th>\n",
       "      <td>1</td>\n",
       "      <td>31</td>\n",
       "      <td>17</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAAAATAAGCTCACCC</th>\n",
       "      <td>19</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAAATCCTGTCGCCCA</th>\n",
       "      <td>81</td>\n",
       "      <td>121</td>\n",
       "      <td>80</td>\n",
       "      <td>94</td>\n",
       "      <td>32</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAATCGGCATACCATG</th>\n",
       "      <td>82</td>\n",
       "      <td>74</td>\n",
       "      <td>121</td>\n",
       "      <td>89</td>\n",
       "      <td>58</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAATGACATTACTGCA</th>\n",
       "      <td>59</td>\n",
       "      <td>76</td>\n",
       "      <td>74</td>\n",
       "      <td>49</td>\n",
       "      <td>85</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "condition             Unsorted Population      Top Sorted Population       \\\n",
       "replicate                            Rep1 Rep2                  Rep1 Rep2   \n",
       "sgRNA Target Sequence                                                       \n",
       "AAAAAAAAAAAATGCATTCT                    1   31                    17   30   \n",
       "AAAAAAAAATAAGCTCACCC                   19   44                    27    0   \n",
       "AAAAAAAATCCTGTCGCCCA                   81  121                    80   94   \n",
       "AAAAAAATCGGCATACCATG                   82   74                   121   89   \n",
       "AAAAAAATGACATTACTGCA                   59   76                    74   49   \n",
       "\n",
       "condition             Bottom Sorted Population       \n",
       "replicate                                 Rep1 Rep2  \n",
       "sgRNA Target Sequence                                \n",
       "AAAAAAAAAAAATGCATTCT                         0   37  \n",
       "AAAAAAAAATAAGCTCACCC                        27   21  \n",
       "AAAAAAAATCCTGTCGCCCA                        32  142  \n",
       "AAAAAAATCGGCATACCATG                        58  197  \n",
       "AAAAAAATGACATTACTGCA                        85  102  "
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_0 = result.reset_index()\n",
    "result_0 = result_0.drop([\"Target Gene Symbol\"], axis=1)\n",
    "result_0 = result_0.set_index(\"sgRNA Target Sequence\")\n",
    "result_0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Control1</th>\n",
       "      <th>Control2</th>\n",
       "      <th>Sorted1</th>\n",
       "      <th>Sorted2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgRNA</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAAAAAAAAAATGCATTCT</th>\n",
       "      <td>Defb34</td>\n",
       "      <td>1</td>\n",
       "      <td>31</td>\n",
       "      <td>17</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAAAATAAGCTCACCC</th>\n",
       "      <td>Mndal</td>\n",
       "      <td>19</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAAATCCTGTCGCCCA</th>\n",
       "      <td>Cox8c</td>\n",
       "      <td>81</td>\n",
       "      <td>121</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAATCGGCATACCATG</th>\n",
       "      <td>Poldip3</td>\n",
       "      <td>82</td>\n",
       "      <td>74</td>\n",
       "      <td>121</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAATGACATTACTGCA</th>\n",
       "      <td>Bcas2</td>\n",
       "      <td>59</td>\n",
       "      <td>76</td>\n",
       "      <td>74</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Gene  Control1  Control2  Sorted1  Sorted2\n",
       "sgRNA                                                              \n",
       "AAAAAAAAAAAATGCATTCT   Defb34         1        31       17       54\n",
       "AAAAAAAAATAAGCTCACCC    Mndal        19        44       27        0\n",
       "AAAAAAAATCCTGTCGCCCA    Cox8c        81       121       80      218\n",
       "AAAAAAATCGGCATACCATG  Poldip3        82        74      121      164\n",
       "AAAAAAATGACATTACTGCA    Bcas2        59        76       74       98"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b_top.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'sgRNA'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2524\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2525\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2526\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'sgRNA'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-173-d195b1748c66>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mb_top\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb_top\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"sgRNA\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mset_index\u001b[0;34m(self, keys, drop, append, inplace, verify_integrity)\u001b[0m\n\u001b[1;32m   3144\u001b[0m                 \u001b[0mnames\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3145\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3146\u001b[0;31m                 \u001b[0mlevel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mframe\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3147\u001b[0m                 \u001b[0mnames\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3148\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mdrop\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2137\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2138\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2139\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2141\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_getitem_column\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2144\u001b[0m         \u001b[0;31m# get column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2145\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2146\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2147\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2148\u001b[0m         \u001b[0;31m# duplicate columns & possible reduce dimensionality\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_get_item_cache\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m   1840\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1841\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1842\u001b[0;31m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1843\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_box_item_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1844\u001b[0m             \u001b[0mcache\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, item, fastpath)\u001b[0m\n\u001b[1;32m   3841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3842\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3843\u001b[0;31m                 \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3844\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3845\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2525\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2526\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2527\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2528\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2529\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'sgRNA'"
     ]
    }
   ],
   "source": [
    "b_top = b_top.set_index([\"sgRNA\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_1 = result_0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgRNA Target Sequence\n",
       "AAAAAAAAAAAATGCATTCT     30\n",
       "AAAAAAAAATAAGCTCACCC      0\n",
       "AAAAAAAATCCTGTCGCCCA     94\n",
       "AAAAAAATCGGCATACCATG     89\n",
       "AAAAAAATGACATTACTGCA     49\n",
       "AAAAAACAACCTATACGAAG     24\n",
       "AAAAAACAGCAAGAAGTGGT      9\n",
       "AAAAAACAGTGGATATCAGA     29\n",
       "AAAAAACCCCCTTAAAGAGT     38\n",
       "AAAAAACGGAATGAGTCTGC     92\n",
       "AAAAAACGTATGGCACGAAG     27\n",
       "AAAAAACTCACAAAAAAGAA      0\n",
       "AAAAAAGTAGCAAGAGCTTT     24\n",
       "AAAAAAGTATATCATACCCC      7\n",
       "AAAAAAGTCCCACAAACAAA     22\n",
       "AAAAAAGTGGAAGACTTCAC     41\n",
       "AAAAAAGTTATTAGCCATGG     33\n",
       "AAAAAATACTCTTCCGTCGA     42\n",
       "AAAAAATGATTAATGACCCG     47\n",
       "AAAAAATGCTTCCACAACTT     93\n",
       "AAAAAATGGCTTCTGGACCG    126\n",
       "AAAAAATGGTGGAATCCATC     84\n",
       "AAAAAATTATGGCCCACGAG     51\n",
       "AAAAACAACAGGGAAGGCCG     33\n",
       "AAAAACACCCTATCAATACA     80\n",
       "AAAAACAGAACACCGAACAT      0\n",
       "AAAAACAGCGAATAGAAACA     18\n",
       "AAAAACAGGATCTCTATATG      0\n",
       "AAAAACAGTTGGCTTACGTG     40\n",
       "AAAAACATGGGCAAATCGTA     50\n",
       "                       ... \n",
       "TTTGTGGATCAGCATTAAGG      0\n",
       "TTTGTGGGCGAAACCCCTGG     60\n",
       "TTTGTGGGCGCTGCTATAGG     59\n",
       "TTTGTGGGCTTACCTACACG     41\n",
       "TTTGTGGTAGTTCTGTTGGT    129\n",
       "TTTGTGTATTCCCGAGTTGG     37\n",
       "TTTGTGTCAGGCTAACCTAG     61\n",
       "TTTGTGTCTTCCAGCAAGCG      9\n",
       "TTTGTGTGAAGAGATACAAT     68\n",
       "TTTGTGTGAGCAATTAAAGG     95\n",
       "TTTGTGTGGAACCGTGACCC     24\n",
       "TTTGTGTTACCTGCATCAGT    129\n",
       "TTTGTTAACTTACGTTTCAT     72\n",
       "TTTGTTAAGATGTTTCAGGG     26\n",
       "TTTGTTACTTTAGATTAGCA     75\n",
       "TTTGTTATCGATGGTCCAAG    128\n",
       "TTTGTTATTACACTTACTGA     91\n",
       "TTTGTTCACGAGGCTTTCCA     67\n",
       "TTTGTTGAACCAAAAAGGGA     12\n",
       "TTTGTTGCAATTGGGGCTTG    140\n",
       "TTTGTTTAGTGCTCAACAAG     93\n",
       "TTTGTTTATTCGACAACTGT    116\n",
       "TTTGTTTGAGGTCCTCAGGT     72\n",
       "TTTGTTTGCAAAGATCACAA     50\n",
       "TTTGTTTGCTCAAGGAGTTG     75\n",
       "TTTGTTTGTTTCAGGCAAAG     91\n",
       "TTTTAGTAGGTCAACGCTGT     36\n",
       "TTTTCGTCGACTAAGTCAAG     85\n",
       "TTTTGGCGACGCATCGCTGG    137\n",
       "TTTTTAGACCTAATTCGCGC      0\n",
       "Name: (Top Sorted Population, Rep2), Length: 79633, dtype: int64"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1[(\"Top Sorted Population\", \"Rep2\")]\n",
    "# b_top[\"Sorted2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_1[(\"Top Sorted Population\",\"Rep2\")] = b_top[\"Sorted2\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_2 = result_1.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA Target Sequence</th>\n",
       "      <th>condition</th>\n",
       "      <th>replicate</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>287869</th>\n",
       "      <td>GCTACAGCAAGTGGGAACGG</td>\n",
       "      <td>Top Sorted Population</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sgRNA Target Sequence              condition replicate  frequency\n",
       "287869  GCTACAGCAAGTGGGAACGG  Top Sorted Population      Rep2         88"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_3 = result_2.melt(id_vars=[\"sgRNA Target Sequence\"], value_name=\"frequency\")\n",
    "result_4 = result_3.loc[result_3[\"condition\"] == \"Top Sorted Population\"]\n",
    "result_5 = result_4.loc[result_4[\"replicate\"] == \"Rep2\"]\n",
    "result_5.loc[result_5[\"sgRNA Target Sequence\"]==\"GCTACAGCAAGTGGGAACGG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_1 = c_guides.loc[(c_guides[\"condition\"] == \"Top Sorted Population\") & (c_guides[\"replicate\"] == \"Rep2\")]\n",
    "\n",
    "c_new = pd.merge(c_1, result_5, on=\"sgRNA Target Sequence\", how=\"left\", suffixes=(\"_old\",\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_new = c_new.drop([\"frequency_old\",\"replicate_old\",\"condition_old\",\"Unnamed: 0\"],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA Target Sequence</th>\n",
       "      <th>mat</th>\n",
       "      <th>Target Gene ID</th>\n",
       "      <th>Target Gene Symbol</th>\n",
       "      <th>Target Transcript</th>\n",
       "      <th>Genomic Sequence</th>\n",
       "      <th>Position of Base After Cut 1 based</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Target Context Sequence</th>\n",
       "      <th>PAM Sequence</th>\n",
       "      <th>Exon Number</th>\n",
       "      <th>Rule Set 2 score</th>\n",
       "      <th>Summary</th>\n",
       "      <th>Description</th>\n",
       "      <th>condition</th>\n",
       "      <th>replicate</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACAGCAAGTGGGAACGG</td>\n",
       "      <td>0</td>\n",
       "      <td>66112</td>\n",
       "      <td>1-Mar</td>\n",
       "      <td>NM_001290273.1</td>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>184810952.0</td>\n",
       "      <td>sense</td>\n",
       "      <td>GGCGGCTACAGCAAGTGGGAACGGTGGCAC</td>\n",
       "      <td>TGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mitochondrial amidoxime reducing component 1</td>\n",
       "      <td>Top Sorted Population</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGAGAAATCAGGACCAATCG</td>\n",
       "      <td>0</td>\n",
       "      <td>66112</td>\n",
       "      <td>1-Mar</td>\n",
       "      <td>NM_001290273.1</td>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>184807131.0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>GGTCAGAGAAATCAGGACCAATCGAGGCTC</td>\n",
       "      <td>AGG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.6506</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mitochondrial amidoxime reducing component 1</td>\n",
       "      <td>Top Sorted Population</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGGCAGAGTGCACTGCCATG</td>\n",
       "      <td>0</td>\n",
       "      <td>66112</td>\n",
       "      <td>1-Mar</td>\n",
       "      <td>NM_001290273.1</td>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>184810878.0</td>\n",
       "      <td>sense</td>\n",
       "      <td>AGCGAGGCAGAGTGCACTGCCATGGGGCTG</td>\n",
       "      <td>GGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mitochondrial amidoxime reducing component 1</td>\n",
       "      <td>Top Sorted Population</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCTCAGTGCAGCTTACACAA</td>\n",
       "      <td>0</td>\n",
       "      <td>66112</td>\n",
       "      <td>1-Mar</td>\n",
       "      <td>NM_001290273.1</td>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>184807073.0</td>\n",
       "      <td>sense</td>\n",
       "      <td>TGACTCTCAGTGCAGCTTACACAAAGGACC</td>\n",
       "      <td>AGG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.7089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mitochondrial amidoxime reducing component 1</td>\n",
       "      <td>Top Sorted Population</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TGTGTTGATAGATCGGACAG</td>\n",
       "      <td>0</td>\n",
       "      <td>72925</td>\n",
       "      <td>1-Mar</td>\n",
       "      <td>NM_175188.4</td>\n",
       "      <td>NC_000074.6</td>\n",
       "      <td>66456168.0</td>\n",
       "      <td>sense</td>\n",
       "      <td>TGTATGTGTTGATAGATCGGACAGCGGAGG</td>\n",
       "      <td>CGG</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.7670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>membrane-associated ring finger (C3HC4) 1</td>\n",
       "      <td>Top Sorted Population</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sgRNA Target Sequence  mat  Target Gene ID Target Gene Symbol  \\\n",
       "0  GCTACAGCAAGTGGGAACGG    0           66112              1-Mar   \n",
       "1  AGAGAAATCAGGACCAATCG    0           66112              1-Mar   \n",
       "2  AGGCAGAGTGCACTGCCATG    0           66112              1-Mar   \n",
       "3  TCTCAGTGCAGCTTACACAA    0           66112              1-Mar   \n",
       "4  TGTGTTGATAGATCGGACAG    0           72925              1-Mar   \n",
       "\n",
       "  Target Transcript Genomic Sequence  Position of Base After Cut 1 based   \\\n",
       "0    NM_001290273.1      NC_000067.6                          184810952.0   \n",
       "1    NM_001290273.1      NC_000067.6                          184807131.0   \n",
       "2    NM_001290273.1      NC_000067.6                          184810878.0   \n",
       "3    NM_001290273.1      NC_000067.6                          184807073.0   \n",
       "4       NM_175188.4      NC_000074.6                           66456168.0   \n",
       "\n",
       "      Strand         Target Context Sequence PAM Sequence  Exon Number  \\\n",
       "0      sense  GGCGGCTACAGCAAGTGGGAACGGTGGCAC          TGG          1.0   \n",
       "1  antisense  GGTCAGAGAAATCAGGACCAATCGAGGCTC          AGG          2.0   \n",
       "2      sense  AGCGAGGCAGAGTGCACTGCCATGGGGCTG          GGG          1.0   \n",
       "3      sense  TGACTCTCAGTGCAGCTTACACAAAGGACC          AGG          2.0   \n",
       "4      sense  TGTATGTGTTGATAGATCGGACAGCGGAGG          CGG          7.0   \n",
       "\n",
       "   Rule Set 2 score Summary                                   Description  \\\n",
       "0            0.6825     NaN  mitochondrial amidoxime reducing component 1   \n",
       "1            0.6506     NaN  mitochondrial amidoxime reducing component 1   \n",
       "2            0.6398     NaN  mitochondrial amidoxime reducing component 1   \n",
       "3            0.7089     NaN  mitochondrial amidoxime reducing component 1   \n",
       "4            0.7670     NaN     membrane-associated ring finger (C3HC4) 1   \n",
       "\n",
       "               condition replicate  frequency  \n",
       "0  Top Sorted Population      Rep2         88  \n",
       "1  Top Sorted Population      Rep2         83  \n",
       "2  Top Sorted Population      Rep2        108  \n",
       "3  Top Sorted Population      Rep2         16  \n",
       "4  Top Sorted Population      Rep2         34  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_mageck_analysis(self, input_dataframe, condition):\n",
    "    \n",
    "    self.mageck_output_path_prefix = \"Mageck\"\n",
    "\n",
    "    # Create mageck input file\n",
    "    status = \"Creating Mageck input file...\"\n",
    "    if condition==\"Top_Sorted\":\n",
    "        mageck_df = create_mageck_input_files(input_dataframe, [\"Unsorted Population\", \"Top Sorted Population\"])\n",
    "    elif condition==\"Bottom_Sorted\":\n",
    "        mageck_df = create_mageck_input_files(input_dataframe, [\"Unsorted Population\", \"Bottom Sorted Population\"])\n",
    "    else:\n",
    "        raise \"Mageck condition not recognized.\"\n",
    "    mageck_path = self.mageck_output_path_prefix+\"_\"+condition+\"_mageck_input_file.txt\"\n",
    "\n",
    "    output_column_values = mageck_df.columns.get_level_values(1)+\":\"+mageck_df.columns.get_level_values(0)\n",
    "    mageck_df.to_csv(mageck_path, sep=\"\\t\", index=True, header=output_column_values)\n",
    "\n",
    "    # Perform mageck test, passing the file path location of the input file\n",
    "    mageck_result_df = self.run_mageck_command(mageck_path, condition)\n",
    "\n",
    "    status = \"Command complete.\"\n",
    "\n",
    "    # Sort by mageck p-value and then write out mageck output file\n",
    "    mageck_result_df = mageck_result_df.sort_values(by=[\"pos|-log(p-value)\"], ascending=False)\n",
    "\n",
    "\n",
    "    status = \"Saving Mageck results to .csv file...\"\n",
    "    mageck_output_path = self.mageck_output_path_prefix+\"_\"+condition+\"_mageck_gene_results.csv\"\n",
    "    mageck_result_df.to_csv(mageck_output_path)\n",
    "\n",
    "    status = \"Mageck complete.\"\n",
    "\n",
    "    return mageck_result_df\n",
    "\n",
    "def run_mageck_command(self, mageck_path, condition):\n",
    "\n",
    "    control_columns = \",\".join(list(map(str,np.arange(self.replicate_number)))) # 0,1 for 2 replicates\n",
    "    test_columns = \",\".join(list(map(str,np.arange(self.replicate_number)+self.replicate_number))) # 2,3 for 2 replicates\n",
    "\n",
    "    command = \"mageck test -k \" + mageck_path + \" -t \"+test_columns+\" -c \"+control_columns+\" -n \" + \\\n",
    "            str(self.mageck_output_path_prefix)+\"_mageck_\" + condition + \\\n",
    "            \" --sort-criteria pos --gene-lfc-method alphamean --control-sgrna \"\\\n",
    "             + str(self.control_file)\n",
    "\n",
    "    status = \"Running Mageck command for %s...\" % condition\n",
    "\n",
    "    os.system(command)\n",
    "\n",
    "    mageck_prefix = self.mageck_output_path_prefix+\"_mageck_\"+condition+\".gene_summary.txt\"\n",
    "    mageck_result_df = pd.read_csv(mageck_prefix, sep=\"\\t\")\n",
    "    mageck_result_df[\"neg|-log(p-value)\"] = mageck_result_df[\"neg|p-value\"].apply(lambda x: -1*math.log(x, 10))\n",
    "    mageck_result_df[\"pos|-log(p-value)\"] = mageck_result_df[\"pos|p-value\"].apply(lambda x: -1*math.log(x, 10))\n",
    "\n",
    "    return mageck_result_df\n",
    "\n",
    "def ratio_test(self, input_dataframe):\n",
    "\n",
    "    status = \"Creating Ratio inputs...\"\n",
    "\n",
    "    bot5_comb = create_mageck_input_files(input_dataframe, [\"Bottom Sorted Population\"])\n",
    "    top5_comb = create_mageck_input_files(input_dataframe, [\"Top Sorted Population\"])\n",
    "\n",
    "    bot5_comb += 1\n",
    "    top5_comb += 1\n",
    "\n",
    "    forQuant = bot5_comb.join(top5_comb)\n",
    "\n",
    "    status = \"Performing quantile normalization...\"\n",
    "\n",
    "    # Add 1 to reads\n",
    "    quantNorm = self.quantileNormalize(forQuant)\n",
    "    quantNorm = quantNorm.rename(columns={\"Bottom Sorted Population\": \"Bottom Quantile Norm\",\n",
    "                                            \"Top Sorted Population\": \"Top Quantile Norm\"})\n",
    "\n",
    "    status = \"Performing geometric mean of quantile normalized values...\"\n",
    "\n",
    "    quantNorm[\"Mean Bottom Quantile Norm\"] = quantNorm[\"Bottom Quantile Norm\"].apply(self.geomMean, axis=1)\n",
    "    quantNorm[\"Mean Top Quantile Norm\"] = quantNorm[\"Top Quantile Norm\"].apply(self.geomMean, axis=1)\n",
    "\n",
    "    status = \"Flattening multi-dimensional index\"\n",
    "\n",
    "    ratios = quantNorm[[\"Mean Bottom Quantile Norm\",\"Mean Top Quantile Norm\"]].copy()\n",
    "    ratios.columns = list(map(lambda column_name: \"\".join(column_name).rstrip(\"\"), ratios.columns))\n",
    "\n",
    "    status = \"Computing ratio of bottom to top...\"\n",
    "\n",
    "    ratios[\"Ratio\"] = ratios[\"Mean Bottom Quantile Norm\"] / ratios[\"Mean Top Quantile Norm\"]\n",
    "\n",
    "    status = \"Computing log2 of ratio...\"\n",
    "\n",
    "    ratios[\"log2(ratio)\"] = ratios[\"Ratio\"].apply(np.log2)\n",
    "\n",
    "    status = \"Computing mean abundance...\"\n",
    "\n",
    "    ratios['mean_abundance'] = forQuant.apply(self.geomMean, axis=1)\n",
    "\n",
    "    status = \"Computing log2(mean abundance)\"\n",
    "\n",
    "    ratios['log_MA'] = ratios['mean_abundance'].apply(np.log2)\n",
    "\n",
    "    ratios = ratios.sort_values(by=['log_MA'], ascending=True)\n",
    "\n",
    "    status = \"Computing binned zscores...\"\n",
    "\n",
    "    bins=3\n",
    "    guidesPerBin = math.ceil(len(ratios)/bins)\n",
    "\n",
    "    for i in range(bins):\n",
    "        start = (i)*guidesPerBin\n",
    "        end = start+guidesPerBin\n",
    "        if end > len(ratios):\n",
    "            end = len(ratios)\n",
    "        guides = ratios[start:end]\n",
    "        guides[\"ZScore\"] = stats.zscore(guides[\"log2(ratio)\"])\n",
    "        ratios.loc[start:end,'ZScore'] = guides.loc[:,'ZScore']\n",
    "\n",
    "    status = \"Outputting guide level file...\"\n",
    "\n",
    "    guides_path = os.path.join(self.output_path, \"Ratio\", self.output_prefix+\"_ratio_guides.csv\")\n",
    "    ratios.to_csv(guides_path)\n",
    "\n",
    "    status = \"Collapsing to gene-level statistics...\"\n",
    "\n",
    "    gene_level_df = ratios.groupby(\"Target Gene Symbol\", as_index=True).agg(\"mean\")\n",
    "    gene_stats = gene_level_df[[\"log2(ratio)\",\"ZScore\"]]\n",
    "    gene_stats = gene_stats.rename(columns={\"log2(ratio)\":\"Mean log2(ratio)\", \"ZScore\":\"Mean ZScore\"})\n",
    "    gene_stats = gene_stats.sort_values(by=\"Mean ZScore\", ascending=False)\n",
    "\n",
    "    status = \"Outputting gene level file...\"\n",
    "\n",
    "    genes_path = os.path.join(self.output_path, \"Ratio\", self.output_prefix+\"_ratio_genes.csv\")\n",
    "    gene_stats.to_csv(genes_path)\n",
    "\n",
    "    return ratios\n",
    "\n",
    "def geomMean(self, x):\n",
    "    product = np.prod(x)\n",
    "    numElem = len(x)\n",
    "    result = product**(1/numElem)\n",
    "    return result\n",
    "\n",
    "def quantileNormalize(self, df_input):\n",
    "    df = df_input.copy()\n",
    "    #compute rank\n",
    "    dic = {}\n",
    "    for col in df:\n",
    "        dic.update({col : sorted(df[col])})\n",
    "    sorted_df = pd.DataFrame(dic)\n",
    "    rank = sorted_df.mean(axis = 1).tolist()\n",
    "    #sort\n",
    "    for col in df:\n",
    "        t = np.searchsorted(np.sort(df[col]), df[col])\n",
    "        df[col] = [rank[i] for i in t]\n",
    "    return df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
